backend/corpora/dataset_processing/upload.py

```diff
@@ -0,0 +1,139 @@
+import logging
+import threading
+import queue
+import requests
+
+from ..common.corpora_orm import DbDatasetProcessingStatus, UploadStatus
+from ..common.entities import Dataset
+from ..common.utils.db_utils import db_session_manager
+from ..common.utils.math_utils import MB
+
+logger = logging.getLogger(__name__)
+
+
+class ProgressTracker:
+    def __init__(self, file_size: int):
+        self.file_size: int = file_size
+        self._progress: int = 0
+        self.progress_lock: threading.Lock = threading.Lock()  # prevent concurrent access of _progress
+        self.stop_updater: threading.Event = threading.Event()  # Stops the update_progress thread
+        self.stop_uploader: threading.Event = threading.Event()  # Stops the uploader threads
+        self.error: queue.Queue = queue.Queue(maxsize=1)  # Track errors
+
+    def progress(self):
+        with self.progress_lock:
+            return self._progress / self.file_size
+
+    def update(self, progress):
+        with self.progress_lock:
+            self._progress += progress
+
+
+def uploader(url: str, local_path: str, tracker: ProgressTracker, chunk_size: int):
+    """
+    Upload the file pointed at by the URL to the local path.
+
+    :param url: The URL of the file to be uploaded.
+    :param local_path: The local name of the file be uploaded
+    :param tracker: Tracks information about the progress of the upload.
+    :return:
+    """
+    try:
+        with requests.get(url, stream=True) as resp:
+            resp.raise_for_status()
+            with open(local_path, "wb") as fp:
+                for chunk in resp.iter_content(chunk_size=chunk_size):
+                    if tracker.stop_uploader.isSet():
+                        logger.debug("Upload ended early!")
```

Comment:
Can you test that this is logged properly? https://docs.python.org/3.4/library/unittest.html#unittest.TestCase.assertLogs

Reply:
done